plugins/vbe/fit.c

```diff
@@ -0,0 +1,302 @@
+	if (!val)
+		return -FITE_NOT_FOUND;
+	*valp = fdt32_to_cpu(*val);
+
+	return 0;
+}
+
+int fit_first_cfg(struct fit_info *fit)
+{
+	int subnode, node;
+
+	node = fdt_path_offset(fit->blob, FIT_CONFIG_PATH);
+	if (node < 0)
+		return -FITE_NO_CONFIG_NODE;
+
+	subnode = fdt_first_subnode(fit->blob, node);
+	if (subnode < 0)
+		return -FITE_NOT_FOUND;
+
+	return subnode;
+}
+
+int fit_next_cfg(struct fit_info *fit, int preb_cfg)
+{
+	int subnode;
+
+	subnode = fdt_next_subnode(fit->blob, preb_cfg);
+	if (subnode < 0)
+		return -FITE_NOT_FOUND;
+
+	return subnode;
+}
+
+const char *fit_cfg_name(struct fit_info *fit, int cfg)
+{
+	return fdt_get_name(fit->blob, cfg, NULL);
+}
+
+const char *fit_cfg_compat_item(struct fit_info *fit, int cfg, int index)
+{
+	return fdt_stringlist_get(fit->blob, cfg, FIT_PROP_COMPATIBLE, index, NULL);
+}
+
+int fit_cfg_img_count(struct fit_info *fit, int cfg, const char *prop_name)
+{
+	int count;
+
+	count = fdt_stringlist_count(fit->blob, cfg, prop_name);
+	if (count < 0)
+		return -FITE_NOT_FOUND;
+
+	return count;
+}
+
+int fit_cfg_img(struct fit_info *fit, int cfg, const char *prop_name, int index)
+{
+	const char *name;
+	int images, image;
+
+	name = fdt_stringlist_get(fit->blob, cfg, prop_name, index, NULL);
+	if (!name)
+		return -FITE_NOT_FOUND;
+
+	images = fdt_path_offset(fit->blob, FIT_IMAGE_PATH);
+	if (images < 0)
+		return -FITE_NO_IMAGES_NODE;
+
+	image = fdt_subnode_offset(fit->blob, images, name);
+	if (image < 0)
+		return -FITE_MISSING_IMAGE;
+
+	return image;
+}
+
+const char *fit_cfg_version(struct fit_info *fit, int cfg)
+{
+	return fdt_getprop(fit->blob, cfg, FIT_VERSION, NULL);
+}
+
+const char *fit_img_name(struct fit_info *fit, int img)
+{
+	return fdt_get_name(fit->blob, img, NULL);
+}
+
+static enum fit_algo_t fit_get_algo(struct fit_info *fit, int node)
+{
+	const char *algo;
+	int i;
+
+	algo = fdt_getprop(fit->blob, node, FIT_PROP_ALGO, NULL);
+	if (!algo)
+		return -FITE_MISSING_ALGO;
+
+	for (i = 0; i < FIT_ALGO_COUNT; i++) {
+		if (!strcmp(fit_algo[i], algo))
+			return i;
+	}
+
+	return -FITE_UNKNOWN_ALGO;
+}
+
+int fit_check_hash(struct fit_info *fit, int node, const char *data, int size)
+{
+	int algo;
+	const char *value;
+	int val_size;
+
+	value = fdt_getprop(fit->blob, node, FIT_PROP_VALUE, &val_size);
+	if (!value)
+		return -FITE_MISSING_VALUE;
+
+	/* Only check the algo after we have found a value */
+	algo = (int)fit_get_algo(fit, node);
+	if (algo < 0)
+		return algo;
+
+	switch (algo) {
+	case FIT_ALGO_CRC32: {
+		unsigned long actual;
+		unsigned long expect;
+
+		if (val_size != 4)
+			return -FITE_INVALID_HASH_SIZE;
+		expect = fdt32_to_cpu(*(fdt32_t *)value);
+		actual = crc32(0, (unsigned char *)data, size);
+		if (expect != actual)
+			return -FITE_HASH_MISMATCH;
+		break;
+	}
+	default:
+		return -FITE_UNKNOWN_ALGO;
+	}
+
+	return 0;
+}
+
+int fit_check_hashes(struct fit_info *fit, int img, const char *data, int size)
+{
+	int node;
+	int ret;
```

Comment:
I actually think I need to break this library out into a separate place, but for now I have not done this.

When I do, I need to follow this style rather than fwupd, avoid glib, etc.

Should I try to do that split-out now? It would involve cloning from the new repo and building it, I suppose.

Reply:
OK so I should keep the library in here for now? If so, should I put it in a different directory for now?

In any case, can I avoid adding glib to it? It would create an API that I don't want

Reply:
> so I should keep the library in here for now

I'm not 100% comfortable shipping libraries in plugin directories; it kinda sets a bad precedent for the next vendor that wants to ship some vendor library but doesn't want to use the existing project style and all the existing GLib and fwupd helper code. @superm1 opinions welcome too.

Reply:
If splitting out to a library that is carried in tree so this can be reused elsewhere is the direction this needs to go, can we consider making it a safer language like rust or go?

Reply:
Well libfdt is written in C (as is fwupd) so at the least we must have a C library. We could add other versions later, of course, but I think C is a good starting point.

As to safety, we should be able to rely on tests to cover that.

Reply:
This isn't going to be popular, but I do think we need to be strict about the requirements here: Either the FIT stuff is useful to other projects and gets included in `libfdt` (and in an upstream release, and included in distros) or this code needs to be in GLib/GObject style. We can't have code that's simultaneously important enough to be used in multiple places, but also not important enough to be part of an existing library or new project.

I told Wacom they couldn't just copy and paste libwacom.cpp into a new plugin, told Synaptics they couldn't just put the existing crufty reference code into fwupd, that USI global couldn't just use their existing windows update code, and most recently Intel they couldn't just import metee and libigsc as source files into fwupd.

I know it seems draconian, but having a consistent codebase is the only way fwupd as a project remains maintainable. If we allow just one vendor to do it then fwupd becomes a dumping ground for legacy code -- and that's not a healthy place to be.

Reply:
Getting code into libfdt, getting a new release, etc takes time.
So maybe the right answer is to open a PR into libfdt with `fit.c` as it exists today, and then to get this plugin merged update fit.c to glib based code.  When a new libfdt releases, there can be a dependency bump in fwupd and all the glib code torn out (or conditionally used depending on the use case).

Reply:
The thing is, I think that a lot of `fit.c` can go away if we use GHashTable and GPtrArray instead.